### PR TITLE
BFD-109: RFC for cursor-based paging

### PR DESCRIPTION
**Pull request by RickHawesUSDS**
This PR only contains the RFC for cursor-based paging. 

https://github.com/CMSgov/beneficiary-fhir-data/blob/rick/cursor-paging-rfc/rfcs/0006-cursor-paging.md

### CODE REVIEW

**Review by whytheplatypus — DISMISSED** [review]

File: rfcs/0006-cursor-paging.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+The latency variability (also known as jitter) of a BFD EOB request leads to several problems: 
+- As mentioned, it requires the BFD client to set long timeout values
+- SLO and SLA latency metrics are difficult to specify because normal operations may take a long time to complete
+- Request retries, usually a good practice in distributed systems, can lead to a very wasteful failure-mode where computationally expensive partial results are repeatedly discarded. 
+- Jitter adversely affects the throughput of a system. 
+
+## Proposed Solution
+The proposal is to replace the current offset paging algorithm with a cursor-based algorithm. 
+Internally, the BFD can implement cursor-based paging with an efficient seek query. 
+From a database perspective, the work done is nearly constant for each request.
```
Comment (karlmdavis) [inline rfcs/0006-cursor-paging.md:31]: Can you please also describe this in terms of the SQL that would be executed, as compared to the current SQL queries? It's not clear to me if you're actually using DB cursors (i.e. <https://www.postgresql.org/docs/9.6/plpgsql-cursors.html>) or if you're just using seek-based paging. If you're using both, it's not clear why you feel the combination is important.

**Review by mackec1 — DISMISSED** [review]
While the RFC was vague in some respects (i.e., possibly a definition of the word 'cursor' that will be used within this RFC), I think the premise can be a most useful tool to improve bfd performance.

There are numerous articles that touch upon this query strategy, especially in environments with large datasets (one of my favorites: https://slack.engineering/evolving-api-pagination-at-slack/). 

I think once folks wrap their heads around how this works in practice, we might come up with possibly a slightly more robust response set. For example, returning the total # of rows that the query would return if no limits were applied, etc.). Clients can then plan accordingly (they'd know that they will need to perform 45 requests to fetch all the data).

**Comment by mjburling** [comment]
In the interest of maintaining a tidy backlog of PRs, I'm proposing that we move forward with _this_ PR by accepting the RFC as-is with a **Deferred** status. I've made some modest changes to the RFC addressing formatting and renaming, but the content is largely as it was when it was originally introduced.

**Review by mackec1 — APPROVED** [review]

**Review by mjburling — APPROVED** [review]

File: rfcs/0006-cursor-paging.md (hunk 1 above)
Comment (mjburling) [reply rfcs/0006-cursor-paging.md]: This was captured in the unresolved questions section of the RFC as we accept this RFC under a _Deferred_ status.

File: rfcs/0006-cursor-paging.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+* RFC PR: 
+* JIRA Ticket(s):
+    
+This RFC changes how the BFD handles requests for explanation of benefits (EOB). 
+Internally, it uses a more efficient database query to fetch EOBs. 
+For partners, it provides a way for BFD clients to request resources with lower timeouts and higher throughput.  
+
+## Motivation
+The latency of the BFD's EOB search is directly related to the number of claims a beneficiary has. 
+Since the number of claims per beneficiary varies widely, a BFD client must set 10 to 30-second timeout on their EOB fetches to accommodate a large number of claims in the result set.
```
[the remark this answers is not in the document]
Comment (mjburling) [reply rfcs/0006-cursor-paging.md]: This was captured in the unresolved questions section of the RFC as we accept this RFC under a _Deferred_ status.

File: rfcs/0006-cursor-paging.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+This RFC changes how the BFD handles requests for explanation of benefits (EOB). 
+Internally, it uses a more efficient database query to fetch EOBs. 
+For partners, it provides a way for BFD clients to request resources with lower timeouts and higher throughput.  
+
+## Motivation
+The latency of the BFD's EOB search is directly related to the number of claims a beneficiary has. 
+Since the number of claims per beneficiary varies widely, a BFD client must set 10 to 30-second timeout on their EOB fetches to accommodate a large number of claims in the result set. 
+
+Paging is the term given to the technique to split a single large result set into many smaller result sets. 
+Since smaller result sets take less processing time, paging should imply lower request latency.
```
[the remark this answers is not in the document]
Comment (mjburling) [reply rfcs/0006-cursor-paging.md]: This was captured in the unresolved questions section of the RFC as we accept this RFC under a _Deferred_ status.

File: rfcs/0006-cursor-paging.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+Paging is the term given to the technique to split a single large result set into many smaller result sets. 
+Since smaller result sets take less processing time, paging should imply lower request latency. 
+The FHIR specification and the BFD API support paging. 
+Unfortunately, the BFD's current paging algorithm is an inefficient implementation of offset paging, where the BFD data server queries for the entire result set and then discards the results outside of the requested page. 
+
+The latency variability (also known as jitter) of a BFD EOB request leads to several problems: 
+- As mentioned, it requires the BFD client to set long timeout values
+- SLO and SLA latency metrics are difficult to specify because normal operations may take a long time to complete
+- Request retries, usually a good practice in distributed systems, can lead to a very wasteful failure-mode where computationally expensive partial results are repeatedly discarded. 
+- Jitter adversely affects the throughput of a system.
```
[the remark this answers is not in the document]
Comment (mjburling) [reply rfcs/0006-cursor-paging.md]: This was captured in the unresolved questions section of the RFC as we accept this RFC under a _Deferred_ status.

**Pull request merged**